pkg/controller/elasticsearch/pdb/roles.go

```diff
@@ -0,0 +1,518 @@
+	for _, pdb := range expectedPDBs {
+		delete(toDelete, pdb.Name)
+		// Ensure that the expected PDB is reconciled.
+		if err := reconcilePDB(ctx, k8sClient, es, pdb); err != nil {
+			return fmt.Errorf("while reconciling role-specific PDB %s: %w", pdb.Name, err)
+		}
+	}
+
+	// Delete unnecessary PDBs
+	for name, pdb := range toDelete {
+		if err := k8sClient.Delete(ctx, &pdb); err != nil {
+			return fmt.Errorf("while deleting role-specific PDB %s: %w", name, err)
+		}
+	}
+
+	return nil
+}
+
+// listAllRoleSpecificPDBs lists all role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func listAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) ([]policyv1.PodDisruptionBudget, error) {
+	// List all PDBs in the namespace with the cluster label
+	var pdbList policyv1.PodDisruptionBudgetList
+	if err := k8sClient.List(ctx, &pdbList, client.InNamespace(es.Namespace), client.MatchingLabels{
+		label.ClusterNameLabelName: es.Name,
+	}); err != nil {
+		return nil, err
+	}
+
+	// Filter only PDBs that are owned by this Elasticsearch controller
+	var roleSpecificPDBs []policyv1.PodDisruptionBudget
+	for _, pdb := range pdbList.Items {
+		// Check if this PDB is owned by the Elasticsearch resource
+		if isOwnedByElasticsearch(pdb, es) {
+			roleSpecificPDBs = append(roleSpecificPDBs, pdb)
+		}
+	}
+	return roleSpecificPDBs, nil
+}
+
+// deleteAllRoleSpecificPDBs deletes all existing role-specific PDBs for the cluster by retrieving
+// all PDBs in the namespace with the cluster label and verifying the owner reference.
+func deleteAllRoleSpecificPDBs(ctx context.Context, k8sClient k8s.Client, es esv1.Elasticsearch) error {
+	// List all PDBs in the namespace with the cluster label
+	var pdbList policyv1.PodDisruptionBudgetList
+	if err := k8sClient.List(ctx, &pdbList, client.InNamespace(es.Namespace), client.MatchingLabels{
+		label.ClusterNameLabelName: es.Name,
+	}); err != nil {
+		return err
+	}
+
+	// Delete only PDBs that are owned by this Elasticsearch controller
+	for _, pdb := range pdbList.Items {
+		// Check if this PDB is owned by the Elasticsearch resource
+		if isOwnedByElasticsearch(pdb, es) {
+			if err := k8sClient.Delete(ctx, &pdb); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// isOwnedByElasticsearch checks if a PDB is owned by the given Elasticsearch resource.
+func isOwnedByElasticsearch(pdb policyv1.PodDisruptionBudget, es esv1.Elasticsearch) bool {
+	for _, ownerRef := range pdb.OwnerReferences {
+		if ownerRef.Controller != nil && *ownerRef.Controller &&
+			ownerRef.APIVersion == esv1.GroupVersion.String() &&
+			ownerRef.Kind == esv1.Kind &&
+			ownerRef.Name == es.Name {
+			return true
+		}
+	}
+	return false
+}
+
+// podDisruptionBudgetName returns the name of the PDB.
+func podDisruptionBudgetName(esName string, role string) string {
```

Comment:
Let's maybe move this somewhere around https://github.com/elastic/cloud-on-k8s/blob/1832f4bbc4e280ffeff5df96fd4ce3bf503899b4/pkg/apis/elasticsearch/v1/name.go#L177 ?